src/app/store/observeMiddleware.ts

```diff
@@ -0,0 +1,76 @@
+import { deepEqual } from 'fast-equals';
+import { Dispatch, isAction, Middleware, MiddlewareAPI } from 'redux';
+import { createAction, createCustomAction, isActionOf } from 'typesafe-actions';
+import { RootState } from './types';
+
+interface StoreObserver<T> {
+  id: string;
+  /**
+   * Function to create "something" which will be used to determine if the side effect should run.
+   * deepEqual is used for the comparison, so use primitives, shallow objects or arrays to keep
+   * performance reasonable.
+   */
+  getObserved: (rootState: RootState) => T;
+  /**
+   * Runs the side effect providing both the previous and current version of the derrived state.
+   */
+  sideEffect: (states: { previous: T; current: T }) => void;
+}
+
+// Need to user a higher order function to get the correct typings and inference, it will now correctly
+// type the value of T based on what the getObserved returns
+export const observe = <T>(storeObserver: StoreObserver<T>) =>
+  createCustomAction('observer/OBSERVE', (storeObserver: StoreObserver<T>) => ({ storeObserver }))(
+    storeObserver,
+  );
+export const unobserve = createAction('observer/UNOBSERVE')<string>();
+export const clearObservers = createAction('observer/CLEAR_OBSERVERS')();
+
+function isObserveAction(action: unknown): action is ReturnType<typeof observe> {
+  return (
+    typeof action === 'object' &&
+    action !== null &&
+    'type' in action &&
+    action.type === 'observer/OBSERVE'
+  );
+}
+
+export function observerMiddleware<D extends Dispatch>(
+  api: MiddlewareAPI<D, RootState>,
+): ReturnType<Middleware> {
+  const observers = new Map<string, StoreObserver<unknown>>();
+  return (next) => (action) => {
+    if (!isAction(action)) {
+      return next(action);
+    }
+
+    if (isObserveAction(action)) {
+      observers.set(action.storeObserver.id, action.storeObserver);
```

Comment:
It would be super easy to export a function doing this if we want to also do that. `observers` could be hoisted up to the module without issue I believe.

Reply:
Yeah might be a nice convenience.

Reply:
Left it out for now, since I haven't had a use for it in the two examples I did.